srcts/src/shiny/shinyapp.ts

```diff
@@ -1053,8 +1054,13 @@ class ShinyApp {
         const $tabContent = getTabContent($tabset);
         let tabsetId = $parentTabset.attr("data-tabsetid");
 
-        const $divTag = $(message.divTag.html);
-        const $liTag = $(message.liTag.html);
+        // Create a virtual element where we'll temporarily hold the rendered
+        // nav controls so we can rewrite some attributes and choose where to
+        // insert the new controls.
+        const $fragLi = $("<div>");
+        await renderContentAsync($fragLi, message.liTag, "afterBegin");
```

Comment:
Note B in the mega-comment below says "renderContent must be called on an element that's attached to the document". Is that going to be a problem here?

Reply:
It didn't seem to be in testing but I'll take another look. The mega comment is quite old; while the insert tab message handler hasn't changed much since it was written,  `renderContent()` certainly has seen a few changes.

The comment was added here https://github.com/rstudio/shiny/commit/91dbb0e77bfe137335e628ba314c6c9c399e2b71 in #1794

Reply:
Actually it seems like `renderContentAsync()` is essentially the same but `renderHtml()` has changed. (Leaving crumbs for myself) 
https://github.com/rstudio/shiny/blob/91dbb0e77bfe137335e628ba314c6c9c399e2b71/srcjs/output_binding_html.js

Reply:
I found a very helpful comment from you @jcheng5 in that PR thread that I'll move up to a top-level comment here.

Reply:
The mega comment also states

```
// D) renderContent has a bug where only position "replace" (the default)
//    uses the jQuery functions, so other positions like "beforeend" will
//    prevent child script tags from running.
```

but one core difference is that `renderHtml()` does now use jQuery for all insertion positions. The change was made in #3630.

Reply:
Oh, I found it! Output elements need to be attached to the DOM, otherwise they aren't bound

https://github.com/rstudio/shiny/blob/d764ea9b4e4f773d7dc5c3bfc2e2741441b10b44/srcts/src/shiny/bind.ts#L317-L320

So we just need to call `bindAll()` again after inserting the nav controls, in case they contain outputs. af6ecdbe4cf63ae1f57ffc71030f3b91a6521b65

Reply:
Here's a fun twist: I don't think this is needed anymore. The issue from #1399 doesn't reproduce if I take out the above line (added in #1402). (This change smells like an indirect fix that was reasonable at the time but might not be needed anymore.)

Reply:
After further reflection, I'm fine with this the way you've written it. I was going to suggest doing all the tag manipulation with a parsed-but-not-rendered-or-inserted `$liTag`, then calling renderContentAsync without doing the appending the `$liTag` as a separate step. And if it were me I'd still maybe try to do it, but it's a pretty marginal improvement in code clarity and no difference in behavior AFAICT. So I'll leave it up to you.